android/src/main/java/com/polidea/blemulator/SimulatedAdapter.java

```diff
@@ -335,6 +364,82 @@ public void onSuccess(DeviceContainer deviceContainer) {
         throw new BleError(BleErrorCode.ServiceNotFound, "Service with id " + serviceIdentifier + " not found", 0);
     }
 
+    @Override
+    public List<Descriptor> descriptorsForDevice(String deviceIdentifier, String serviceUUID, String characteristicUUID) throws BleError {
+        if (knownPeripherals.get(deviceIdentifier) == null) {
+            throw new BleError(BleErrorCode.DeviceNotFound, "Device unknown", 0);
+        }
+
+        if (!knownPeripherals.get(deviceIdentifier).isConnected()) {
+            throw new BleError(BleErrorCode.DeviceNotConnected, "Device not connected", 0);
+        }
+
+        if (knownPeripherals.get(deviceIdentifier).getServices() == null) {
+            throw new BleError(BleErrorCode.ServicesNotDiscovered, "Discovery not done on this device", 0);
+        }
+
+        List<CharacteristicContainer> characteristicContainers = knownPeripherals.get(deviceIdentifier)
+                .getCharacteristicContainersIndexedByServiceUuids()
+                .get(serviceUUID);
+
+        List<Descriptor> descriptors = null;
+
+        for (CharacteristicContainer characteristicContainer : characteristicContainers) {
+            String checkedUuid = characteristicContainer.getCharacteristic().getUuid().toString().toUpperCase();
+            if (checkedUuid.equals(characteristicUUID.toUpperCase())) {
+                descriptors = characteristicContainer.getDescriptors();
+                break;
+            }
+        }
+
+        return descriptors;
+    }
+
+    @Override
+    public List<Descriptor> descriptorsForService(int serviceIdentifier, String characteristicUUID) throws BleError {
+        for (DeviceContainer deviceContainer : knownPeripherals.values()) {
+            if (deviceContainer.getServices() != null) {
+                for (Service service : deviceContainer.getServices()) {
+                    if (service.getId() == serviceIdentifier) {
+                        if (!deviceContainer.isConnected()) {
+                            throw new BleError(BleErrorCode.DeviceNotConnected, "Device not connected", 0);
```

Comment:
What if the user would like to get descriptors for other simulated device that is actually connected? Is it possible?

Reply:
Discussed in person: it would have to be done on the FlutterBleLib side.